### DISCUSSION TITLE
System.Collections.HashHelper.GetPrime(int)

### DISCUSSION THREAD

**Discussion by kefyru** [dotnet/runtime#69495] [General]
Is there any exposed api for this function in dotnet?
Is it possible to make this class public?

**Comment by svick** [comment]
What is the justification for making it public?

**Reply by kefyru** [reply]
It is useful function...
Why reinvent the wheel and produce libraries if it is already implemented.
Besides, it does not contain anything dangerous.

**Reply by kefyru** [reply]
I think every internal static class that contains some useful pure functions should be exposed as public.
Especially since many of them use intrinsic optimisations.

**Reply by huoyaoyuan** [reply]
Every public function has a maintenance cost. We can't change or remove it since we are the BCL. We are responsible to keep them correct and fast. We need to properly handle all the possible inputs and corner cases.

If you think it is useful, open an API suggestion.

**Reply by kefyru** [reply]
May be it can be done like with ```CollectionsMarshal.GetValueRefOrAddDefault()```

```cs
class CollectionsMarshal {
   ...
   public int GetPrime(int num) => System.Collections.HashHelper.GetPrime(int);
}
```

**Reply by huoyaoyuan** [reply]
It's nothing about collection. It's indeed math.

**Reply by danmoseley** [reply]
As others said, there has to be a good reason to add new public API. We have to maintain for maybe decades, we have to document, it's something else that shows up on intellisense and as cognitive load, etc, or conversely isn't found. If we want to change it, we now could break people: maybe we would end up making a new private method just for collections.
Also in this specific case it's not a general purpose method. It won't find just any prime. It's not fast enough for all purposes etc. To have a good public API you need to understand needs - what range? What performance? Also, what are callers trying to do - what other API do they need to get the job done? They would likely be more successful if they used a specialized library.
Those are just some of the considerations.
You might find it interesting to watch some of the API reviews as you will see issues like this discussed. Linking to a random one as I am not sure how to link to the series: https://m.youtube.com/watch?v=1t_a9fbB3jY

**Reply by danmoseley** [reply]
Perhaps someone can recommend a math library where you could get this functionality?
